lib/TextDecoder-impl.js

```diff
@@ -0,0 +1,45 @@
+"use strict";
+
+const { labelToName, isSupported, decode } = require("./whatwg-encoding");
+
+exports.implementation = class TextDecoderImpl {
+  constructor(globalObject, constructorArgs = []) {
+    const label = constructorArgs[0];
+    const options = constructorArgs[1] || {};
```

Comment:
No need to be so defensive. The Web IDL will do most of this for you automatically.

```suggestion
  constructor(globalObject, [label, options = {}]) {
```